lib/gc/Dialect/CPURuntime/Transforms/CPURuntimeToLLVM.cpp

```diff
@@ -109,6 +109,78 @@ class PrintfRewriter : public ConvertOpToLLVMPattern<PrintfOp> {
+class AlignedAllocRewriterBase : public ConvertOpToLLVMPattern<OpType> {
+public:
+  using ConvertOpToLLVMPattern<OpType>::ConvertOpToLLVMPattern;
+
+  LogicalResult
+  matchAndRewrite(OpType runtimeOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    Operation *op = runtimeOp;
+    MLIRContext *context = op->getContext();
+    auto moduleOp = op->getParentOfType<ModuleOp>();
+    auto loc = op->getLoc();
+    MemRefType memRefType =
+        dyn_cast<MemRefType>(op->getResults().front().getType());
+    mlir::Type llvmIntPtr = IntegerType::get(
+        context, this->getTypeConverter()->getPointerBitwidth(0));
+    mlir::Type i8Ptr = LLVM::LLVMPointerType::get(context);
+    auto allocFunc = getOrDefineFunction(
+        moduleOp, loc, rewriter, allocFuncName,
+        LLVM::LLVMFunctionType::get(i8Ptr, {llvmIntPtr}, /*isVarArg*/ false));
+
+    auto operands = adaptor.getOperands();
+    SmallVector<Value, 4> shape;
+    SmallVector<Value, 4> strides;
+    Value sizeBytes;
+    this->getMemRefDescriptorSizes(loc, memRefType, operands, rewriter, shape,
+                                   strides, sizeBytes);
+
+    Type elementPtrType = this->getElementPtrType(memRefType);
+    SmallVector<Value, 1> appendFormatArgs = {sizeBytes};
+    LLVM::CallOp allocater =
+        rewriter.create<LLVM::CallOp>(loc, allocFunc, appendFormatArgs);
+    Value allocatedPtr = allocater.getResult();
+    allocatedPtr =
+        rewriter.create<LLVM::BitcastOp>(loc, elementPtrType, allocatedPtr);
+
+    Value alignedPtr = allocatedPtr;
+
+    auto memRefDescriptor = this->createMemRefDescriptor(
+        loc, memRefType, allocatedPtr, alignedPtr, shape, strides, rewriter);
+
+    rewriter.replaceOp(op, {memRefDescriptor});
+    return success();
+  }
+};
+
+template <typename OpType, typename OpAdaptor, const char *deallocFuncName>
+class AlignedFreeRewriterBase : public ConvertOpToLLVMPattern<OpType> {
+public:
+  using ConvertOpToLLVMPattern<OpType>::ConvertOpToLLVMPattern;
+  LogicalResult
+  matchAndRewrite(OpType runtimeOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    Operation *op = runtimeOp;
+    MLIRContext *context = op->getContext();
+    auto moduleOp = op->getParentOfType<ModuleOp>();
+    auto loc = op->getLoc();
+    mlir::Type i8Ptr = LLVM::LLVMPointerType::get(context);
+    mlir::Type llvmVoidType = LLVM::LLVMVoidType::get(context);
+    auto deallocFunc = getOrDefineFunction(
+        moduleOp, loc, rewriter, deallocFuncName,
+        LLVM::LLVMFunctionType::get(llvmVoidType, {i8Ptr}, /*isVarArg*/ false));
+    Value pointer =
+        MemRefDescriptor(adaptor.getMemref()).allocatedPtr(rewriter, loc);
+    auto casted = rewriter.create<LLVM::BitcastOp>(loc, i8Ptr, pointer);
+    SmallVector<Value, 1> appendFormatArgs = {casted};
+    rewriter.create<LLVM::CallOp>(loc, deallocFunc, appendFormatArgs);
+    rewriter.eraseOp(op);
```

Comment:
Any reason for not using `replaceOp()`?

Reply:
because we do not have output for deallocFunc, so there's no dependent users need to be replaced.

Reply:
I see, thanks for the info.